File: iosApp/deadly/Core/Database/DAOs/SyncOutboxDAO.swift
import Foundation
import GRDB

struct SyncOutboxDAO: Sendable {
    let database: AppDatabase

    /// Idempotent enqueue. If a pending entry for (kind, refId) already
    /// exists, this no-ops — the flusher will read the row's current
    /// state at push time so the final state wins.
    func enqueue(kind: String, refId: String) throws {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        try database.write { db in
            try db.execute(sql: """
                INSERT OR IGNORE INTO sync_outbox (kind, refId, createdAt, attemptCount)
                VALUES (?, ?, ?, 0)
            """, arguments: [kind, refId, now])
        }
    }

    func fetchPending(kind: String) throws -> [SyncOutboxRecord] {
        try database.read { db in
            try SyncOutboxRecord
                .filter(Column("kind") == kind)
                .order(Column("createdAt"))
                .fetchAll(db)
        }
    }

    func delete(id: Int64) throws {
        _ = try database.write { db in
            try SyncOutboxRecord.deleteOne(db, key: id)
        }
    }

    func recordFailure(id: Int64, error: String) throws {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        try database.write { db in
            try db.execute(sql: """
                UPDATE sync_outbox
                   SET lastAttemptAt = ?, attemptCount = attemptCount + 1, lastError = ?
                 WHERE id = ?
            """, arguments: [now, error, id])
        }
    }

    func pendingCount(kind: String) throws -> Int {
        try database.read { db in
            try SyncOutboxRecord.filter(Column("kind") == kind).fetchCount(db)
        }
    }
}

File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/dao/SyncOutboxDao.kt
package com.grateful.deadly.core.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.grateful.deadly.core.database.entities.SyncOutboxEntity

@Dao
interface SyncOutboxDao {
    /** Idempotent enqueue. If a pending row for (kind, refId) exists, this no-ops. */
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun enqueue(entry: SyncOutboxEntity): Long

    @Query("SELECT * FROM sync_outbox WHERE kind = :kind ORDER BY createdAt ASC")
    suspend fun fetchPending(kind: String): List<SyncOutboxEntity>

    @Query("DELETE FROM sync_outbox WHERE id = :id")
    suspend fun delete(id: Long)

    @Query("""
        UPDATE sync_outbox
           SET lastAttemptAt = :now, attemptCount = attemptCount + 1, lastError = :error
         WHERE id = :id
    """)
    suspend fun recordFailure(id: Long, now: Long, error: String)

    @Query("SELECT COUNT(*) FROM sync_outbox WHERE kind = :kind")
    suspend fun pendingCount(kind: String): Int
}
